core-foundation/src/array.rs

```diff
@@ -12,31 +12,56 @@
 pub use core_foundation_sys::array::*;
 pub use core_foundation_sys::base::{CFIndex, CFRelease};
 use core_foundation_sys::base::{CFTypeRef, kCFAllocatorDefault};
+use base::CFType;
 use libc::c_void;
 use std::mem;
+use std::marker::PhantomData;
 
 use base::{CFIndexConvertible, TCFType, CFRange};
 
 /// A heterogeneous immutable array.
-pub struct CFArray(CFArrayRef);
+pub struct CFArray<T = *const c_void>(CFArrayRef, PhantomData<T>);
 
-impl Drop for CFArray {
+/// A trait describing how to convert from the stored *const c_void to the desired T
+pub trait FromVoid {
+    fn from_void(x: *const c_void) -> Self;
+}
+
+impl FromVoid for u32 {
+    fn from_void(x: *const c_void) -> u32 {
+        x as usize as u32
+    }
+}
+
+impl FromVoid for *const c_void {
+    fn from_void(x: *const c_void) -> *const c_void {
+        x
+    }
+}
+
+impl FromVoid for CFType {
+    fn from_void(x: *const c_void) -> CFType {
+        unsafe { TCFType::wrap_under_get_rule(mem::transmute(x)) }
```

Comment:
This is unsound, one can pass `0xdeadbeef as *const c_void` or any other arbitrary and trigger UB.